common/lru/blob_lru.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2022 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package lru
+
+import (
+	"math"
+	"sync"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/hashicorp/golang-lru/simplelru"
+)
+
+// SizeConstrainedLRU is a wrapper around simplelru.LRU. The simplelru.LRU is capable
+// of item-count constraints, but is not capable of enforcing a byte-size constraint,
+// hence this wrapper.
+type SizeConstrainedLRU struct {
+	size    uint64
+	maxSize uint64
+	lru     *simplelru.LRU
+	lock    sync.RWMutex
+}
+
+// NewSizeConstraiedLRU creates a new SizeConstrainedLRU.
+func NewSizeConstraiedLRU(max uint64) *SizeConstrainedLRU {
+	lru, err := simplelru.NewLRU(math.MaxInt, nil)
+	if err != nil {
+		panic(err)
+	}
+	return &SizeConstrainedLRU{
+		size:    0,
+		maxSize: max,
+		lru:     lru,
+	}
+}
+
+// Set adds a value to the cache.  Returns true if an eviction occurred.
+func (c *SizeConstrainedLRU) Set(key []byte, value []byte) (evicted bool) {
+	return c.Add(string(key), string(value))
+}
+
+// Add adds a value to the cache.  Returns true if an eviction occurred.
+func (c *SizeConstrainedLRU) Add(key string, value string) (evicted bool) {
```

Comment:
This won't work correctly if you are overwriting a key with a different value. Actually, it will also double count the size if you add the same value for an existing item.

Reply:
Good point!

Reply:
I added docs to highlight that this cache assumes a mapping between keys and values. In our case, the key will be the hash of the value (code). 
The case of adding the same item multiple times has been fixed.